heron/instance/src/java/com/twitter/heron/instance/AbstractOutputCollector.java

```diff
@@ -93,23 +94,30 @@ public long getTotalTuplesEmitted() {
   }
 
   protected HeronTuples.HeronDataTuple.Builder initTupleBuilder(String streamId,
-                                                                List<Object> tuple) {
+                                                                List<Object> tuple,
+                                                                Integer taskId) {
     // Start construct the data tuple
     HeronTuples.HeronDataTuple.Builder builder = HeronTuples.HeronDataTuple.newBuilder();
 
     // set the key. This is mostly ignored
     builder.setKey(0);
 
     List<Integer> customGroupingTargetTaskIds = null;
-    if (!helper.isCustomGroupingEmpty()) {
+    if (taskId != null) {
+      // TODO: somehow assert that the input stream of the downstream bolt was configured
+      // with directGrouping
+
+      customGroupingTargetTaskIds = new ArrayList<>();
```

Comment:
can we use a separate `directGroupingTragetTaskIds` to handle direct grouping cases? Also instead of using `if... else if`, can we separate the handling of direct grouping and custom grouping?